File: Improve/Classes/Provider/AppGivensProvider.m
//
//  AppGivensProvider.m
//  ImproveUnitTests
//
//  Created by PanHongxi on 10/27/21.
//  Copyright © 2021 Mind Blown Apps, LLC. All rights reserved.
//
#import <UIKit/UIKit.h>

#import <CoreTelephony/CTTelephonyNetworkInfo.h>
#import <CoreTelephony/CTCarrier.h>
#import <sys/sysctl.h>
#import <sys/types.h>

#import "AppGivensProvider.h"
#import "IMPDecisionModel.h"
#import "IMPConstants.h"

@implementation IMPDeviceInfo

- (instancetype)initWithModel:(NSString *)model version:(double)version {
    if(self = [super init]) {
        _model = model;
        _version = [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.3lf", version]];
    }
    return self;
}

@end

@implementation AppGivensProvider

+ (instancetype)shared {
    static AppGivensProvider *instance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[AppGivensProvider alloc] init];
    });
    return instance;
}

static NSString * PersistPrefix = @"ai.improve";

static NSString * const kCountryKey = @"$country";
NSString * const kLanguageKey = @"$lang";
static NSString * const kTimezoneKey = @"$tz";
static NSString * const kCarrierKey = @"$carrier";
static NSString * const kDeviceKey = @"$device";
static NSString * const kDeviceVersionKey = @"$devicev";
static NSString * const kOSKey = @"$os";
static NSString * const kOSVersionKey = @"$osv";
static NSString * const kScreenPixelsKey = @"$pixels";
static NSString * const kAppKey = @"$app";
static NSString * const kAppVersionKey = @"$appv";
static NSString * const kImproveVersionKey = @"$sdkv";
static NSString * const kWeekDayKey = @"$weekday";
static NSString * const kSinceMidnightKey = @"$time";
static NSString * const kSinceSessionStartKey = @"$runtime";
static NSString * const kSinceBornKey = @"$day";
static NSString * const kDecisionCountKey = @"$d"; // number of decisions for this model
static NSString * const kRewardsKey = @"$r";  // total rewards for this model
static NSString * const kRewardsPerDecision = @"$r/d";
static NSString * const kDecisionsPerDay = @"$d/day";

// persistent key

// The first time an AppGivensProvider is created on this device.
static NSString * const kBornTimeKey = @"ai.improve.born_time";

// When the AppGivensProvider instance is created.
// If there is some way to get the true app launch
// time in Android or iOS we could use that instead.
NSString * const kSessionStartTimeKey = @"ai.improve.session_start_time";

static NSString * const kDefaultsDecisionCountKey = @"ai.improve.decision_count-%@";

static NSString * const kDefaultsModelRewardsKey = @"ai.improve.rewards-%@";

NSString * const kGivensModelNameMessages = @"messages";
NSString * const kGivensModelNameThemes = @"themes";
NSString * const kGivensModelNameStories = @"stories";
NSString * const kGivensModelNameSongs = @"songs";

static double sLastSessionStartTime;

- (instancetype)init {
    if(self = [super init]) {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
            
            // cache last session start time before it is overwritten in NSUserDefaults
            // by current session start time
            sLastSessionStartTime = [defaults doubleForKey:kSessionStartTimeKey];
            
            // set born time
            if([defaults objectForKey:kBornTimeKey] == nil) {
                double bornTime = [[NSDate date] timeIntervalSince1970];
                [defaults setDouble:bornTime forKey:kBornTimeKey];
            }
            
            // set session start time
            double sessionStartTime = [[NSDate date] timeIntervalSince1970];
            [defaults setDouble:sessionStartTime forKey:kSessionStartTimeKey];
        });
    }
    return self;
}

- (NSDictionary<NSString *, id> *)givensForModel:(IMPDecisionModel *)decisionModel givens:(NSDictionary *)givens_ {
    NSMutableDictionary *givens = [[NSMutableDictionary alloc] init];
    givens[kCountryKey] = [self country];
    givens[kLanguageKey] = [self language];
    givens[kTimezoneKey] = @([self timezone]);
    givens[kCarrierKey] = [self carrier];
    givens[kOSKey] = [self os];
    givens[kOSVersionKey] = [self osVersion];
    givens[kAppKey] = [self app];
    givens[kAppVersionKey] = [self appVersion];
    givens[kImproveVersionKey] = [self improveVersion:kIMPVersion];
    givens[kScreenPixelsKey] = @([self screenPixels]);
    givens[kWeekDayKey] = [self weekDay];
    givens[kSinceMidnightKey] = [self sinceMidnight];
    givens[kSinceBornKey] = [self sinceBornDecimalNumber];
    givens[kSinceSessionStartKey] = [self sinceSessionStart];
    givens[kDecisionCountKey] = @([self decisionCount:decisionModel.modelName]);
    givens[kRewardsKey] = [self rewardOfModelDecimalNumber:decisionModel.modelName];
    givens[kRewardsPerDecision] = [self rewardsPerDecision:decisionModel.modelName];
    givens[kDecisionsPerDay] = [self decisionsPerDay:decisionModel.modelName];
    
    IMPDeviceInfo *deviceInfo = [self deviceInfo];
    givens[kDeviceKey] = deviceInfo.model;
    givens[kDeviceVersionKey] = deviceInfo.version;
    
    // When getGivens is called, increment decision count value by 1
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *decisionCountKey = [NSString stringWithFormat:kDefaultsDecisionCountKey, decisionModel.modelName];
    NSInteger curDecisionCount = [defaults integerForKey:decisionCountKey];
    [defaults setInteger:curDecisionCount+1 forKey:decisionCountKey];

    // If keys in givens_ overlap with keys in AppGivensProvider, keys in givens_ win
    [givens addEntriesFromDictionary:givens_];
    
    return givens;
}

//  two letter code
- (NSString *)country {
    return [[NSLocale currentLocale] objectForKey:NSLocaleCountryCode];
}

//  two letter code
- (NSString *)language {
    NSString *lang = [[NSLocale preferredLanguages] firstObject];
    NSDictionary *languageDic = [NSLocale componentsFromLocaleIdentifier:lang];
    return [languageDic objectForKey:NSLocaleLanguageCode];
}

// numeric GMT offset
- (NSInteger)timezone {
    return [[NSTimeZone localTimeZone] secondsFromGMT] / 3600;
}

// When there are multiple carriers, the first non-empty carrierName is used.
- (NSString *)carrier {
    CTTelephonyNetworkInfo *info = [[CTTelephonyNetworkInfo alloc] init];
    NSDictionary<NSString *, CTCarrier *> *carriers = info.serviceSubscriberCellularProviders;
    for(NSString *key in carriers) {
        CTCarrier *carrier = carriers[key];
        if(carrier.carrierName.length > 0) {
            return carrier.carrierName;
        }
    }
    return nil;
}

- (NSString *)os {
#if TARGET_OS_OSX
    return @"macOS";
#elif TARGET_OS_TV
    return @"tvOS";
#elif TARGET_OS_MACCATALYST
    return @"macOS";
#elif TARGET_OS_WATCH
    return @"watchOS";
#else
    return @"iOS";
#endif
}

- (NSDecimalNumber *)osVersion {
    NSOperatingSystemVersion systemVersion = [[NSProcessInfo processInfo] operatingSystemVersion];
    double t = systemVersion.majorVersion + systemVersion.minorVersion/1000.0 + systemVersion.patchVersion/1000000.0;
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", t]];
}

- (IMPDeviceInfo *)deviceInfo {
    NSString *platform = [AppGivensProvider getPlatformString];
    return [self parseDeviceInfo:platform];
}

- (NSString *)app {
    NSString *name = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleDisplayName"];
    if([name length] <= 0) {
        name = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleName"];
    }
    return name == nil ? @"" : name;
}

// When running the unit test, appVersion is nil and can't be put in the dict.
// I'm making it empty string instead of nil to avoid the problem for now.
- (NSDecimalNumber *)appVersion {
    NSString *appVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
    double t = [self versionToNumber:appVersion];
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", t]];
}

- (NSDecimalNumber *)improveVersion:(NSString *)version {
    double t = [self versionToNumber:version];
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", t]];
    
}

- (double)versionToNumber:(NSString *)version {
    NSArray<NSString *> *versionArray = [version componentsSeparatedByString:@"."];
    
    int major = 0;
    int minor = 0;
    int build = 0;
    
    if([versionArray count] == 1) {
        major = [versionArray[0] intValue];
    } else if([versionArray count] == 2) {
        major = [versionArray[0] intValue];
        minor = [versionArray[1] intValue];
    } else if([versionArray count] >= 3) {
        major = [versionArray[0] intValue];
        minor = [versionArray[1] intValue];
        build = [versionArray[2] intValue];
    }
    return major + minor / 1000.0 + build / 1000000.0;
}

- (NSUInteger)screenPixels {
    CGFloat scale = [[UIScreen mainScreen] scale];
    CGRect rect = [[UIScreen mainScreen] bounds];
    return rect.size.width * scale * rect.size.height * scale;
}

- (NSDecimalNumber *)weekDay {
    NSDate *now = [NSDate date];
    NSCalendar *gregorian = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
    [gregorian setFirstWeekday:2];
    NSUInteger weekday = [gregorian ordinalityOfUnit:NSCalendarUnitWeekday inUnit:NSCalendarUnitWeekOfMonth forDate:now];
    double sinceMidnight = [now timeIntervalSinceDate:[gregorian startOfDayForDate:now]];
    double t = weekday + sinceMidnight / (24 * 3600);
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.5lf", t]];
}

- (NSDecimalNumber *)sinceMidnight {
    NSCalendar *gregorian = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
    NSDate *now = [NSDate date];
    NSDate *startOfDay = [gregorian startOfDayForDate:now];
    double t = ([now timeIntervalSinceDate:startOfDay]) / 86400.0;
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", t]];
}

// When the AppGivensProvider instance is created.
// If there is some way to get the true app launch
// time in Android or iOS we could use that instead.
- (NSDecimalNumber *)sinceSessionStart {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    double sessionStartTime = [defaults doubleForKey:kSessionStartTimeKey];
    double t = ([[NSDate date] timeIntervalSince1970] - sessionStartTime) / 86400.0;
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", t]];
}

// Born time is the first time an AppGivensProvider is created on this device.
- (NSDecimalNumber *)sinceBornDecimalNumber {
    double t = [self sinceBorn];
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", t]];
}

- (double)sinceBorn {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    double bornTime = [defaults doubleForKey:kBornTimeKey];
    return ([[NSDate date] timeIntervalSince1970] - bornTime) / 86400.0;
}

//  It’s the number of times a givens is provided
// 0 is returned for the first decision
- (NSUInteger)decisionCount:(NSString *)modelName {
    NSString *decisionCountKey = [NSString stringWithFormat:kDefaultsDecisionCountKey, modelName];
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    return [defaults integerForKey:decisionCountKey];
}

// https://www.theiphonewiki.com/wiki/Models#iPhone
// Check the link above to see what platform might look like.
// Pay attention to filed 'identifier'
- (IMPDeviceInfo *)parseDeviceInfo:(NSString *)platform {
    if(platform.length <= 0) {
        return [[IMPDeviceInfo alloc] initWithModel:@"unknown" version:0];
    }
    
    if([platform isEqualToString:@"i386"]) {
        return [[IMPDeviceInfo alloc] initWithModel:@"Simulator" version:0];
    } else if([platform isEqualToString:@"x86_64"]) {
        return [[IMPDeviceInfo alloc] initWithModel:@"Simulator" version:0];
    }

    NSUInteger lastLetterIndex = 0;
    for(NSUInteger i = 0; i < platform.length; ++i) {
        unichar c = [platform characterAtIndex:i];
        if(!((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z'))) {
            lastLetterIndex = i;
            break;
        }
    }
    if(lastLetterIndex == 0) {
        return [[IMPDeviceInfo alloc] initWithModel:platform version:0];
    }
    
    NSString *model = [platform substringToIndex:lastLetterIndex];
    
    NSString *versionString = [platform substringFromIndex:lastLetterIndex];
    NSArray<NSString *> *versionArray = [versionString componentsSeparatedByString:@","];
    if(versionArray.count != 2) {
        return [[IMPDeviceInfo alloc] initWithModel:model version:0];
    }
    
    int major = [versionArray[0] intValue];
    int minor = [versionArray[1] intValue];
    
    return [[IMPDeviceInfo alloc] initWithModel:model version:major+minor/1000.0];
}

+ (NSString *)getPlatformString {
#if !TARGET_OS_OSX
    const char *sysctl_name = "hw.machine";
#else
    const char *sysctl_name = "hw.model";
#endif
    size_t size;
    sysctlbyname(sysctl_name, NULL, &size, NULL, 0);
    char *machine = malloc(size);
    sysctlbyname(sysctl_name, machine, &size, NULL, 0);
    NSString *platform = [NSString stringWithUTF8String:machine];
    free(machine);
    return platform;
}

+ (void)addReward:(double)reward forModel:(NSString *)modelName {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *key = [NSString stringWithFormat:kDefaultsModelRewardsKey, modelName];
    double curValue = [defaults doubleForKey:key];
    [defaults setDouble:(curValue+reward) forKey:key];
}

// Total rewards for this model
- (NSDecimalNumber *)rewardOfModelDecimalNumber:(NSString *)modelName {
    double reward = [self rewardOfModel:modelName];
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", reward]];
}

- (double)rewardOfModel:(NSString *)modelName {
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    NSString *key = [NSString stringWithFormat:kDefaultsModelRewardsKey, modelName];
    return [defaults doubleForKey:key];
}

- (NSDecimalNumber *)rewardsPerDecision:(NSString *)modelName {
    double rewards = [self rewardOfModel:modelName];
    NSUInteger decisions = [self decisionCount:modelName];
    if(decisions == 0) {
        return [NSDecimalNumber decimalNumberWithString:@"0"];
    } else {
        return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", rewards / decisions]];
    }
}

- (NSDecimalNumber *)decisionsPerDay:(NSString *)modelName {
    NSUInteger decisions = [self decisionCount:modelName];
    double days = [self sinceBorn];
    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%.6lf", decisions / days]];
}

@end

File: Sources/ImproveAI/AppGivensProvider.swift
//
//  AppGivensProvider.swift
//  
//
//  Created by Hongxi Pan on 2022/12/7.
//

import UIKit
import Foundation
import CoreTelephony

struct GivensKey {
    static let context = "context"
    static let country = "country"
    static let lang = "lang"
    static let tz = "tz"
    static let carrier = "carrier"
    static let device = "device"
    static let os = "os"
    static let pixels = "pixels"
    static let app = "app"
    static let weekday = "weekday"
    static let time = "time"
    static let runtime = "runtime"
    static let day = "day"
    static let decisionNumberOfModel = "d"
    static let rewardsOfModel = "r"
    static let rewardPerDecision = "r/d"
    static let decisionsPerDay = "d/day"
}

fileprivate let SecondsPerDay: Double = 86400

fileprivate let ModelRewardsKey = "ai.improve.rewards-%@"

fileprivate let BornTimeKey = "ai.improve.born_time"

fileprivate let DecisionNumberKey = "ai.improve.decision_count-%@"

public struct AppGivensProvider : GivensProvider {
    
    public static let shared = AppGivensProvider()
    
    // Can't get true app launch time. Make do with this...
    let sessionStartTime = Date().timeIntervalSince1970
    
    init() {
        // set born time
        let bornTime = UserDefaults.standard.object(forKey: BornTimeKey)
        if bornTime == nil {
            UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: BornTimeKey)
        }
    }
    
    public func givens(forModel model: DecisionModel, context: Any?) -> [String : Any] {
        var result: [String : Any] = [:]
        
        result[GivensKey.context] = givens
        result[GivensKey.country] = country()
        result[GivensKey.lang] = language()
        result[GivensKey.tz] = timezone()
        result[GivensKey.carrier] = carrier()
        result[GivensKey.device] = deviceIdentifier()
        result[GivensKey.os] = os()
        result[GivensKey.pixels] = pixels()
        result[GivensKey.app] = app()
        result[GivensKey.weekday] = weekday().toDecimal()
        result[GivensKey.time] = fractionalDay().toDecimal()
        result[GivensKey.runtime] = runtime().toDecimal()
        result[GivensKey.day] = day().toDecimal()
        result[GivensKey.decisionNumberOfModel] = decisionNumber(modelName: model.modelName)
        result[GivensKey.rewardsOfModel] = rewardsOfModel(modelName: model.modelName).toDecimal()
        result[GivensKey.rewardPerDecision] = rewardPerDecision(modelName: model.modelName).toDecimal()
        result[GivensKey.decisionsPerDay] = decisionsPerDay(modelName: model.modelName).toDecimal()
        
        // increment decision number
        incrementDecisionNumberOfModel(modelName: model.modelName)
        
        return result
    }
}



extension AppGivensProvider {
    // two letter code (AU)
    func country() -> String? {
        return Locale.current.regionCode?.uppercased()
    }
    
    // two letter code (EN)
    func language() -> String? {
        return Locale.current.languageCode?.uppercased()
    }
    
    // numeric GMT offset
    func timezone() -> Int {
        return TimeZone.current.secondsFromGMT() / 3600
    }
    
    // cell network
    func carrier() -> String? {
        let info = CTTelephonyNetworkInfo()
        if let carriers = info.serviceSubscriberCellularProviders {
            for (_, carrier) in carriers {
                if let carrierName = carrier.carrierName, !carrierName.isEmpty {
                    return carrierName
                }
            }
        }
        return nil
    }
    
    // device model including version (iPad11,6)
    func deviceIdentifier() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        return machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
    }
    
    // os including version (iOS 16.0.2)
    func os() -> String {
        let systemVersion = ProcessInfo.processInfo.operatingSystemVersion
        return "iOS \(systemVersion.majorVersion).\(systemVersion.minorVersion).\(systemVersion.patchVersion)"
    }
    
    // screen width x screen height
    func pixels() -> Int {
        let scale = UIScreen.main.scale
        let rect = UIScreen.main.bounds
        return Int(rect.size.width * scale * rect.size.height * scale)
    }
    
    // app name and version (#Mindful 6.1.1)
    func app() -> String? {
        let bundle = Bundle.main
        let appName = bundle.object(forInfoDictionaryKey: "CFBundleDisplayName") as? String ?? (bundle.object(forInfoDictionaryKey: "CFBundleName") as? String ?? "")
        let appVersion = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? ""
        return "\(appName) \(appVersion)"
    }
    
    // (ISO 8601, monday==1.0, sunday==7.0) plus fractional part of day
    func weekday() -> Double {
        let now = Date()
        let calendar = Calendar(identifier: .iso8601)
        let weekday = calendar.ordinality(of: .weekday, in: .weekOfMonth, for: now)!
        return Double(weekday) + fractionalDay()
    }
    
    // fractional day since midnight
    func fractionalDay() -> Double {
        let now = Date()
        let calendar = Calendar(identifier: .iso8601)
        let midnight = calendar.startOfDay(for: now)
        let second = calendar.dateComponents([.second], from: midnight, to: now).second!
        return Double(second) / SecondsPerDay
    }
    
    // fractional days since session start
    func runtime() -> Double {
        return (Date().timeIntervalSince1970 - self.sessionStartTime) / SecondsPerDay
    }
    
    // fractional days since born
    func day() -> Double {
        if let bornTime = UserDefaults.standard.object(forKey: BornTimeKey) as? Double {
            return (Date().timeIntervalSince1970 - bornTime) / SecondsPerDay
        }
        return 0
    }
    
    // the number of decisions for this model
    func decisionNumber(modelName: String) -> Int {
        let key = String(format: DecisionNumberKey, modelName)
        return UserDefaults.standard.integer(forKey: key)
    }
    
    // total rewards for this model
    func rewardsOfModel(modelName: String) -> Double {
        let key = String(format: ModelRewardsKey, modelName)
        return UserDefaults.standard.double(forKey: key)
    }
    
    // total rewards / decisions number
    func rewardPerDecision(modelName: String) -> Double {
        let decisionNumber = decisionNumber(modelName: modelName)
        let rewards = rewardsOfModel(modelName: modelName)
        return decisionNumber == 0 ? 0 : rewards / Double(decisionNumber)
    }
    
    // decision number / day
    func decisionsPerDay(modelName: String) -> Double {
        let decisionNumber = decisionNumber(modelName: modelName)
        let day = day()
        return Double(decisionNumber) / day
    }
}

extension AppGivensProvider {
    static func addReward(_ reward: Double, forModel modelName: String) {
        let key = String(format: ModelRewardsKey, modelName)
        let curTotalReward = UserDefaults.standard.double(forKey: key)
        UserDefaults.standard.set(curTotalReward + reward, forKey: key)
    }
    
    func incrementDecisionNumberOfModel(modelName: String) {
        let key = String(format: DecisionNumberKey, modelName)
        let curNum = UserDefaults.standard.integer(forKey: key)
        UserDefaults.standard.setValue(curNum+1, forKey: key)
    }
}

extension Double {
    func toDecimal() -> NSDecimalNumber {
        return NSDecimalNumber(string: String(format: "%.6lf", self))
    }
}
